File: app/vue/src/store/types/sales.ts
export interface SalesAgency {
  id: number
  project: number
  name: string
  is_direct_managed: boolean
  business_number: string
  ceo_name: string
  phone: string
  order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface SalesTeam {
  id: number
  agency: number
  agency_name?: string
  parent: number | null
  parent_name?: string | null
  name: string
  order: number
  is_active: boolean
  members_count?: number
  created_at: string
}

export type SalesDuty = '1' | '2' | '3' | '4' | '5' // 1:상담사, 2:팀장, 3:본부장, 4:총괄본부장, 5:지원/기타
export type SalesPersonStatus = '1' | '2' | '3' // 1:재직, 2:휴직, 3:해촉
export type TaxType = '1' | '2' | '3' | '4' // 1:3.3%사업소득, 2:근로소득, 3:사업자, 4:기타
export type SalesDocType = '1' | '2' | '3' | '4' | '5' | '9' // 1:등본, 2:통장, 3:신분증, 4:위촉계약서, 5:각종서약서, 9:기타

export interface SalesPersonDocument {
  id: number
  sales_person: number
  sales_person_name?: string
  doc_type: SalesDocType
  doc_type_display?: string
  title: string
  file: string
  file_name: string
  file_type: string
  file_size: number | null
  is_verified: boolean
  verified_at: string | null
  verified_by: number | null
  verified_by_name?: string | null
  uploader: number | null
  uploader_name?: string | null
  created_at: string
  updated_at: string
}

export interface SalesPerson {
  id: number
  team: number
  team_name?: string
  agency_name?: string
  user: number | null
  name: string
  duty: SalesDuty
  duty_display?: string
  status: SalesPersonStatus
  status_display?: string
  phone: string
  id_number: string
  tax_type: TaxType
  tax_type_display?: string
  bank_name: string
  account_number: string
  account_holder: string
  join_date: string | null
  quit_date: string | null
  notes: string
  documents_count?: number
  documents?: SalesPersonDocument[]
  created_at: string
  updated_at: string
}

export interface CommissionPolicy {
  id: number
  project: number
  order_group: number | null
  order_group_name?: string
  unit_type: number | null
  unit_type_name?: string
  name: string
  agent_fee: number
  leader_fee: number
  director_fee: number
  agency_fee: number
  pay_condition: '1' | '2' | '3' | '4'
  pay_condition_display?: string
  start_date: string
  end_date: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface ContractSalesAgent {
  id: number
  contract: number
  contract_serial?: string
  contractor_name?: string
  order_group_name?: string
  unit_type_name?: string
  unit_info?: string
  agency?: number | null
  agency_name?: string
  is_direct_managed?: boolean
  sales_person: number | null
  sales_person_name?: string
  team: number | null
  team_name?: string
  policy: number | null
  policy_name?: string
  contract_date: string | null
  mgm_name: string
  mgm_phone: string
  mgm_fee: number
  note: string
  is_settlement_approved: boolean
  approval_note?: string
  approved_by?: number | null
  approved_by_name?: string | null
  approved_at?: string | null
  is_settled?: boolean
  settled_period_title?: string | null
  created_at: string
  updated_at: string
}

export interface PayoutContractDetail {
  id: number
  payout: number
  contract: number
  contract_serial?: string
  contractor_name?: string
  role_type: 'agent' | 'leader' | 'director' | 'mgm'
  role_type_display?: string
  unit_fee: number
}

export interface CommissionPayout {
  id: number
  period: number
  sales_person: number
  sales_person_name?: string
  duty_display?: string
  team_name?: string
  base_pay: number
  contract_count: number
  commission_amount: number
  bonus_amount: number
  deduction_amount: number
  gross_amount: number
  income_tax: number
  local_income_tax: number
  total_tax: number
  net_amount: number
  pay_status: '1' | '2' | '3' | '4' // 1:대기, 2:승인, 3:지급완료, 4:지급보류
  pay_status_display?: string
  paid_date: string | null
  bank_name: string
  account_number: string
  account_holder: string
  note: string
  contract_details?: PayoutContractDetail[]
  created_at: string
  updated_at: string
}

export interface SettlementPeriod {
  id: number
  project: number
  title: string
  start_date: string
  end_date: string
  payout_date: string | null
  status: '1' | '2' | '3' // 1:작성중, 2:확정, 3:지급완료
  status_display?: string
  total_contracts: number
  total_gross_amount: number
  total_tax_amount: number
  total_net_amount: number
  agency_fee_total?: number
  billing_supply_price?: number
  billing_vat?: number
  billing_total_amount?: number
  payout_count?: number
  agency_payout_count?: number
  created_by: number | null
  created_at: string
  updated_at: string
}

export interface AgencyPayoutContractDetail {
  id: number
  payout: number
  contract: number
  contract_serial?: string
  contractor_name?: string
  unit_fee: number
}

export interface AgencyPayout {
  id: number
  period: number
  agency: number
  agency_name?: string
  is_direct_managed?: boolean
  contract_count: number
  agency_fee_sum: number
  vat_amount: number
  total_amount: number
  pay_status: '1' | '2' | '3' | '4'
  pay_status_display?: string
  paid_date: string | null
  business_number: string
  bank_name: string
  account_number: string
  account_holder: string
  note: string
  contract_details?: AgencyPayoutContractDetail[]
  created_at: string
  updated_at: string
}

export interface CommissionClawback {
  id: number
  contract: number
  contract_serial?: string
  sales_person: number
  sales_person_name?: string
  amount: number
  reason: string
  is_settled: boolean
  settled_payout: number | null
  created_at: string
}

File: app/django/apiV1/serializers/sales.py
from rest_framework import serializers
from sales.models import (
    SalesAgency, SalesTeam, SalesPerson, SalesPersonDocument, CommissionPolicy,
    ContractSalesAgent, SettlementPeriod, CommissionPayout,
    PayoutContractDetail, CommissionClawback, AgencyPayout, AgencyPayoutContractDetail,
)


class SalesAgencySerializer(serializers.ModelSerializer):
    class Meta:
        model = SalesAgency
        fields = (
            'id', 'project', 'name', 'is_direct_managed',
            'business_number', 'ceo_name', 'phone', 'order',
            'is_active', 'created_at', 'updated_at'
        )


class SalesTeamSerializer(serializers.ModelSerializer):
    agency_name = serializers.ReadOnlyField(source='agency.name')
    parent_name = serializers.ReadOnlyField(source='parent.name')
    members_count = serializers.SerializerMethodField()

    class Meta:
        model = SalesTeam
        fields = (
            'id', 'agency', 'agency_name', 'parent', 'parent_name',
            'name', 'order', 'is_active', 'members_count', 'created_at'
        )

    def get_members_count(self, obj):
        return obj.members.filter(status='1').count()


class SalesPersonDocumentSerializer(serializers.ModelSerializer):
    sales_person_name = serializers.ReadOnlyField(source='sales_person.name')
    doc_type_display = serializers.CharField(source='get_doc_type_display', read_only=True)
    verified_by_name = serializers.ReadOnlyField(source='verified_by.username')
    uploader_name = serializers.ReadOnlyField(source='uploader.username')

    class Meta:
        model = SalesPersonDocument
        fields = (
            'id', 'sales_person', 'sales_person_name', 'doc_type', 'doc_type_display',
            'title', 'file', 'file_name', 'file_type', 'file_size',
            'is_verified', 'verified_at', 'verified_by', 'verified_by_name',
            'uploader', 'uploader_name', 'created_at', 'updated_at'
        )
        read_only_fields = ('file_name', 'file_type', 'file_size', 'uploader', 'created_at', 'updated_at')


class SalesPersonSerializer(serializers.ModelSerializer):
    team_name = serializers.ReadOnlyField(source='team.name')
    agency_name = serializers.ReadOnlyField(source='team.agency.name')
    duty_display = serializers.CharField(source='get_duty_display', read_only=True)
    status_display = serializers.CharField(source='get_status_display', read_only=True)
    tax_type_display = serializers.CharField(source='get_tax_type_display', read_only=True)
    documents_count = serializers.SerializerMethodField()
    documents = SalesPersonDocumentSerializer(many=True, read_only=True)

    class Meta:
        model = SalesPerson
        fields = (
            'id', 'team', 'team_name', 'agency_name', 'user',
            'name', 'duty', 'duty_display', 'status', 'status_display',
            'phone', 'id_number', 'tax_type', 'tax_type_display',
            'bank_name', 'account_number', 'account_holder',
            'join_date', 'quit_date', 'notes', 'documents_count', 'documents',
            'created_at', 'updated_at'
        )

    def get_documents_count(self, obj):
        return obj.documents.count()


class CommissionPolicySerializer(serializers.ModelSerializer):
    unit_type_name = serializers.ReadOnlyField(source='unit_type.name')
    order_group_name = serializers.ReadOnlyField(source='order_group.name')
    pay_condition_display = serializers.CharField(source='get_pay_condition_display', read_only=True)

    class Meta:
        model = CommissionPolicy
        fields = (
            'id', 'project', 'order_group', 'order_group_name',
            'unit_type', 'unit_type_name', 'name',
            'agent_fee', 'leader_fee', 'director_fee', 'agency_fee',
            'pay_condition', 'pay_condition_display',
            'start_date', 'end_date', 'is_active',
            'created_at', 'updated_at'
        )


class ContractSalesAgentSerializer(serializers.ModelSerializer):
    contract_serial = serializers.ReadOnlyField(source='contract.serial_number')
    contractor_name = serializers.ReadOnlyField(source='contract.contractor.name')
    order_group_name = serializers.ReadOnlyField(source='contract.order_group.name')
    unit_type_name = serializers.ReadOnlyField(source='contract.unit_type.name')
    unit_info = serializers.SerializerMethodField()
    agency_name = serializers.ReadOnlyField(source='agency.name')
    is_direct_managed = serializers.ReadOnlyField(source='agency.is_direct_managed')
    sales_person_name = serializers.ReadOnlyField(source='sales_person.name')
    team_name = serializers.ReadOnlyField(source='team.name')
    policy_name = serializers.ReadOnlyField(source='policy.name')
    approved_by_name = serializers.ReadOnlyField(source='approved_by.username')
    is_settled = serializers.SerializerMethodField()
    settled_period_title = serializers.SerializerMethodField()

    class Meta:
        model = ContractSalesAgent
        fields = (
            'id', 'contract', 'contract_serial', 'contractor_name',
            'order_group_name', 'unit_type_name', 'unit_info',
            'agency', 'agency_name', 'is_direct_managed',
            'sales_person', 'sales_person_name', 'team', 'team_name',
            'policy', 'policy_name', 'contract_date',
            'mgm_name', 'mgm_phone', 'mgm_fee', 'note',
            'is_settlement_approved', 'approval_note',
            'approved_by', 'approved_by_name', 'approved_at',
            'is_settled', 'settled_period_title',
            'created_at', 'updated_at'
        )

    def validate(self, attrs):
        agency = attrs.get('agency') or (self.instance.agency if self.instance else None)
        sales_person = attrs.get('sales_person') or (self.instance.sales_person if self.instance else None)
        if not agency and not sales_person:
            raise serializers.ValidationError('외주 대행사 또는 담당 영업직원(상담사) 중 하나는 필수 입력해야 합니다.')
        return attrs

    def get_unit_info(self, obj):
        if obj.contract and obj.contract.key_unit and hasattr(obj.contract.key_unit, 'houseunit'):
            hu = obj.contract.key_unit.houseunit
            return f'{hu.building_unit.name}동 {hu.name}호'
        return ''

    def get_is_settled(self, obj):
        # 직영 또는 외주 PayoutContractDetail 존재 여부
        return (
            PayoutContractDetail.objects.filter(contract=obj.contract).exists() or
            AgencyPayoutContractDetail.objects.filter(contract=obj.contract).exists()
        )

    def get_settled_period_title(self, obj):
        pcd = PayoutContractDetail.objects.filter(contract=obj.contract).select_related('payout__period').first()
        if pcd:
            return pcd.payout.period.title
        apcd = AgencyPayoutContractDetail.objects.filter(contract=obj.contract).select_related('payout__period').first()
        if apcd:
            return apcd.payout.period.title
        return None


class PayoutContractDetailSerializer(serializers.ModelSerializer):
    contract_serial = serializers.ReadOnlyField(source='contract.serial_number')
    contractor_name = serializers.ReadOnlyField(source='contract.contractor.name')
    role_type_display = serializers.CharField(source='get_role_type_display', read_only=True)

    class Meta:
        model = PayoutContractDetail
        fields = (
            'id', 'payout', 'contract', 'contract_serial',
            'contractor_name', 'role_type', 'role_type_display', 'unit_fee'
        )


class CommissionPayoutSerializer(serializers.ModelSerializer):
    sales_person_name = serializers.ReadOnlyField(source='sales_person.name')
    duty_display = serializers.CharField(source='sales_person.get_duty_display', read_only=True)
    team_name = serializers.ReadOnlyField(source='sales_person.team.name')
    pay_status_display = serializers.CharField(source='get_pay_status_display', read_only=True)
    contract_details = PayoutContractDetailSerializer(many=True, read_only=True)

    class Meta:
        model = CommissionPayout
        fields = (
            'id', 'period', 'sales_person', 'sales_person_name', 'duty_display', 'team_name',
            'base_pay', 'contract_count', 'commission_amount', 'bonus_amount',
            'deduction_amount', 'gross_amount', 'income_tax', 'local_income_tax',
            'total_tax', 'net_amount', 'pay_status', 'pay_status_display',
            'paid_date', 'bank_name', 'account_number', 'account_holder',
            'note', 'contract_details', 'created_at', 'updated_at'
        )


class SettlementPeriodSerializer(serializers.ModelSerializer):
    status_display = serializers.CharField(source='get_status_display', read_only=True)
    payout_count = serializers.SerializerMethodField()
    agency_payout_count = serializers.SerializerMethodField()

    class Meta:
        model = SettlementPeriod
        fields = (
            'id', 'project', 'title', 'start_date', 'end_date', 'payout_date',
            'status', 'status_display', 'total_contracts',
            'total_gross_amount', 'total_tax_amount', 'total_net_amount',
            'agency_fee_total', 'billing_supply_price', 'billing_vat', 'billing_total_amount',
            'payout_count', 'agency_payout_count', 'created_by', 'created_at', 'updated_at'
        )

    def get_payout_count(self, obj):
        return obj.payouts.count()

    def get_agency_payout_count(self, obj):
        return obj.agency_payouts.count()


class CommissionClawbackSerializer(serializers.ModelSerializer):
    contract_serial = serializers.ReadOnlyField(source='contract.serial_number')
    sales_person_name = serializers.ReadOnlyField(source='sales_person.name')

    class Meta:
        model = CommissionClawback
        fields = (
            'id', 'contract', 'contract_serial', 'sales_person', 'sales_person_name',
            'amount', 'reason', 'is_settled', 'settled_payout', 'created_at'
        )


class AgencyPayoutContractDetailSerializer(serializers.ModelSerializer):
    contract_serial = serializers.ReadOnlyField(source='contract.serial_number')
    contractor_name = serializers.ReadOnlyField(source='contract.contractor.name')

    class Meta:
        model = AgencyPayoutContractDetail
        fields = (
            'id', 'payout', 'contract', 'contract_serial', 'contractor_name', 'unit_fee'
        )


class AgencyPayoutSerializer(serializers.ModelSerializer):
    agency_name = serializers.ReadOnlyField(source='agency.name')
    is_direct_managed = serializers.ReadOnlyField(source='agency.is_direct_managed')
    pay_status_display = serializers.CharField(source='get_pay_status_display', read_only=True)
    contract_details = AgencyPayoutContractDetailSerializer(many=True, read_only=True)

    class Meta:
        model = AgencyPayout
        fields = (
            'id', 'period', 'agency', 'agency_name', 'is_direct_managed',
            'contract_count', 'agency_fee_sum', 'vat_amount', 'total_amount',
            'pay_status', 'pay_status_display', 'paid_date',
            'business_number', 'bank_name', 'account_number', 'account_holder',
            'note', 'contract_details', 'created_at', 'updated_at'
        )
